Add steps to sign included qemu and notarize the built pkg

anjannath opened the pull request:
<!--
Thanks for sending a pull request!

Please make sure you've read our contributing guidelines and how to submit a pull request (https://github.com/containers/podman/blob/main/CONTRIBUTING.md#submitting-pull-requests).

In case you're only changing docs, make sure to prefix the pull-request title with "[CI:DOCS]". That will prevent functional tests from running and save time and energy.

Finally, be sure to sign commits with your real name. Since by opening
a PR you already have commits, you can add signatures if needed with
something like `git commit -s --amend`.
-->

- This adds the missing signing of the qemu binary included in the pkg
- adds makefile target `notarize` to perform notarization of the built pkg installer

#### To test generation of signed and notarized pkg artifact run

```
$ make ARCH=<amd64 | aarch64> CODESIGN_IDENTITY=<ID> PRODUCTSIGN_IDENTITY=<ID> NOTARIZE_USERNAME=<appleID> NOTARIZE_PASSWORD=<appleID-password> notarize
```

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes, please follow the Kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
None

```

anjannath commented:
@benoitf This handles the signing for the `qemu` binary and the needed notarization, however i am not able to test it for lack of a key

ashley-cui commented on contrib/pkginstaller/README.md:
```diff
@@ -5,6 +5,9 @@ $ make ARCH=<amd64 | aarch64> NO_CODESIGN=1 pkginstaller
 $ make ARCH=<amd64 | aarch64> CODESIGN_IDENTITY=<ID> PRODUCTSIGN_IDENTITY=<ID> pkginstaller
+
+# or to prepare a signed and notarized pkg for release
+$ make ARCH=<amd64 | aarch64> CODESIGN_IDENTITY=<ID> PRODUCTSIGN_IDENTITY=<ID> NOTARIZE_USERNAME=<appleID> NOTARIZE_PASSWORD=<appleID-password> notarize
```
Looking at my notes for how I usually sign and notarize, I only have a DevAppID, an Username, and a AppleID password. I assume the `NOTARIZE_USERNAME` is my username, and the `NOTARIZE_PASSWORD` is the apple ID product password (looks like xxxx-xxxx-xxxx-xxxx). I also think the DevAppID is`CODESIGN_IDENTITY`. What would be the `PRODUCTSIGN_IDENTITY`?

anjannath replied:
can you try to use the DevAppID also for the `PRODUCTSIGN_IDENTITY` i think DevAppID can be used for it if it supports signing installers as well

if the DevAppID doesn't support signing installers then you have to download new certs with installer signing capabilities from the apple developer portal

(https://www.hexnode.com/mobile-device-management/help/how-to-sign-macos-pkg-files-for-deployment-with-hexnode-mdm/)

ashley-cui replied:
Tested it, looks like I need a installer signing identity as well. Going to see if I can generate one today and try it out again.

anjannath commented on contrib/pkginstaller/package.sh:
```diff
@@ -29,16 +33,24 @@ function sign() {
+  # need to remove any extended attributes, otherwise codesign complains:
+  # qemu-system-aarch64: resource fork, Finder information, or similar detritus not allowed
+  xattr -cr "${qemuBinDir}/qemu-system-${arch}"
+  codesign --sign "${CODESIGN_IDENTITY}" --force --entitlements "${BASEDIR}/hvf.entitlements" "${qemuBinDir}/qemu-system-${arch}"
```
(we might need to sign a lot more of the included libraries and binaries inside `qemu/lib, qemu/bin` but want to know before signing everything if only signing the `qemu-system-*` binary passes the apple notarization checks)

anjannath replied:
we do need to sign more, https://github.com/containers/podman/pull/15105#pullrequestreview-1059212224

TomSweeneyRedHat commented:
LGTM, but way out of my wheelhouse.  I'll lean on @ashley-cui for this one.

ashley-cui commented on contrib/pkginstaller/Makefile:
```diff
@@ -45,6 +47,12 @@ package_root: get_gvproxy get_qemu
 	cd $(PACKAGE_DIR) && ./package.sh ..
 
+_notarize: pkginstaller
+	xcrun notarytool submit --apple-id $(NOTARIZE_USERNAME) --password $(NOTARIZE_PASSWORD) -f json --wait out/$(PKG_ARCHIVE_NAME)
```
Alright, I ran into two errors: One: 
```
xcrun notarytool submit --apple-id $(NOTARIZE_USERNAME) --password $(NOTARIZE_PASSWORD) -f json --wait out/$(PKG_ARCHIVE_NAME)`
Error: out must be a zip archive (.zip), flat installer package (.pkg), or UDIF disk image (.dmg)
```

I then re-ran with `PKG_ARCHIVE_NAME=podman-installer-macos-aarch64.pkg` (That seemed to be the pkg name according the to makefile). I'd expect this line to automatically grab PKG_ARCHIVE_NAME and not require me to specify it.

Second, notarytool requires a --team-id flag as well. 
```
xcrun notarytool submit --apple-id $(NOTARIZE_USERNAME) --password $(NOTARIZE_PASSWORD) -f json --wait out/$(PKG_ARCHIVE_NAME)`
Must provide all app-specific password authentication arguments (--apple-id, --password, --team-id). You may leave '--password' unspecified on initial invocation to receive a secure prompt.
```
```suggestion
	xcrun notarytool submit --apple-id $(NOTARIZE_USERNAME) --password $(NOTARIZE_PASSWORD)  --team-id=$(NOTARIZE_TEAM) -f json --wait out/$(PKG_ARCHIVE_NAME)
```

anjannath replied:
Nice, will look at the pkg name issue and update the script as suggested, so it did pass notarisation and apple didn't complain about missing signature on some of the files?

anjannath commented:
Updated the PR as suggested, fixed the pkg name issue, it should've been `PGK_NAME` instead of `PKG_ARCHIVE_NAME` which i forgot to update earlier, `PKG_ARCHIVE_NAME` was not defined in the Makefile, should be good now.

ashley-cui reviewed (commented):
Alright, notarytool now works correctly and submits it, but the notarization fails because we need to sign not only the qemu binary but every dependent binary: Here's a short snippit of the errors that notarytool complains about:
```
{
  "logFormatVersion": 1,
  "jobId": "redacted",
  "status": "Invalid",
  "statusSummary": "Archive contains critical validation errors",
  "statusCode": 4000,
  "archiveFilename": "podman-installer-macos-aarch64.pkg",
  "uploadDate": "2022-08-02T18:01:26.698Z",
  "sha256": "redacted",
  "ticketContents": null,
  "issues": [
    {
      "severity": "error",
      "code": null,
      "path": "podman-installer-macos-aarch64.pkg/podman.pkg Contents/Payload/opt/podman/qemu/bin/vde_autolink",
      "message": "The binary is not signed with a valid Developer ID certificate.",
      "docUrl": null,
      "architecture": "arm64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "podman-installer-macos-aarch64.pkg/podman.pkg Contents/Payload/opt/podman/qemu/bin/vde_autolink",
      "message": "The signature does not include a secure timestamp.",
      "docUrl": null,
      "architecture": "arm64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "podman-installer-macos-aarch64.pkg/podman.pkg Contents/Payload/opt/podman/qemu/bin/vde_autolink",
      "message": "The executable does not have the hardened runtime enabled.",
      "docUrl": null,
      "architecture": "arm64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "podman-installer-macos-aarch64.pkg/podman.pkg Contents/Payload/opt/podman/qemu/bin/msgconv",
      "message": "The binary is not signed with a valid Developer ID certificate.",
      "docUrl": null,
      "architecture": "arm64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "podman-installer-macos-aarch64.pkg/podman.pkg Contents/Payload/opt/podman/qemu/bin/msgconv",
      "message": "The signature does not include a secure timestamp.",
      "docUrl": null,
      "architecture": "arm64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "podman-installer-macos-aarch64.pkg/podman.pkg Contents/Payload/opt/podman/qemu/bin/msgconv",
      "message": "The executable does not have the hardened runtime enabled.",
      "docUrl": null,
      "architecture": "arm64"
    },
```
(This goes on for ~3000 more lines..) 
Looks like its everything in Contents/Payload/opt/podman/qemu/bin and Contents/Payload/opt/podman/qemu/lib needs to be signed in order to be notarized.

anjannath commented:
right, i suspected it'd need everything to be signed but wanted to be a bit conservative before taking that step

so now the script is updated to sign almost everything (anything that has executable permissions, named as '*.dylib' and not a symlink) inside `Contents/Payload/opt/podman/qemu/{bin,lib}`

please give it another try

ashley-cui commented:
Signed and notarize worked for me! Stapled and validated too. Thanks @anjannath! 

/lgtm
/hold

benoitf commented:
are you able to publish the signed pkg somewhere so people can experiment it as well ?

ashley-cui commented on contrib/pkginstaller/package.sh:
```diff
@@ -29,16 +33,35 @@ function sign() {
+  # sign the qemu-system-* binary
+  # need to remove any extended attributes, otherwise codesign complains:
+  # qemu-system-aarch64: resource fork, Finder information, or similar detritus not allowed
+  xattr -cr "${qemuBinDir}/qemu-system-${arch}"
```
Oh, one more review, I've only been testing with aarch64, but this fails with ARCH=amd64, since the amd64 qemu binary is names qemu-system-x6_64

anjannath replied:
ah right, i forgot about this, thanks

anjannath replied:
@ashley-cui update, PTALA

ashley-cui commented:
Oops. One more review, since I forgot to test with amd64. 

@benoitf, github won't allow me to attach .pkg files to PR's, I'll put them in the slack channel, once the amd64 version is fixed :)

ashley-cui commented:
/lgtm cancel

benoitf commented:
@ashley-cui slack works for me (on github you can also zip/tar the file so it can be attached but it needs to be < 25MB)

ashley-cui commented:
/lgtm 

Thanks so much @anjannath :)

ashley-cui commented:
/approve
/hold cancel
